Add support for displaying human-readable breadcrumb labels instead of IDs

FrankeLukas opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Currently breadcrumb labels are just the raw urls, not really user readable. (E.g. the course ID instead of the name.)

### Description
To properly solve this issue, routes need to be re-factored to use children instead of absolute paths. This is a lot of work, so the PR here just starts on a smaller scale (completely porting System Administration, but only parts of the Course Administration). Yet this PR is still pretty big, so I suggest **reviewing the code commit by commit**.

How routes would ideally look using children (and do on this PR):
```
course management
├───create
├───course id
│   ├───update
│   ├───scores
│   ├───exercises
│   │   ├───create
│   │   ├───id
│   │   └───...
│   ├───lectures
│   │   ├───create
│   │   ├───id
│   │   └───...
```
How they more or less look on `develop`:
```
course management
├───create
├───course id
├───course id───update
├───course id───scores
├───course id───exercises
├───course id───exercises───create
├───course id───exercises───id
├───course id───lectures
├───course id───lectures───create
├───course id───lectures───id
├───...
```

The reason fixing this is taking so long is that several components do not use resolvers, but implemented their own assumptions about the routes they are in, so those need to be updated as well. Then there is also the problem with routes being inconsistent and - even more annoyingly - spread out over a lot of files which makes it really hard to get an overview. Lazily loaded routes also can't access routes outside of their root component, which makes sense but means I had to put in a workaround to manually define breadcrumbs for such cases. (This will mostly be used in follow-up PRs since I didn't touch most of the lazily loaded routes.)

### Steps for Testing

1. Log in to Artemis
2. Test all buttons that move you to another page from the Course or System Administration
3. Check that the breadcrumb labels in the System Administration paths are human readable
4. Check that all other breadcrumbs are either unchanged (raw url) or readable
5. Check that the crumbs point to the correct paths

### Screenshots

Breadcrumbs now use human readable labels in several places:
![grafik](https://user-images.githubusercontent.com/72132281/100517039-45207380-3188-11eb-8596-1313fec60fa8.png)
![grafik](https://user-images.githubusercontent.com/72132281/100517059-697c5000-3188-11eb-8d4e-b8f2d2d98fb4.png)

A lot of components still use the old behaviour until their routes have been refactored:
![grafik](https://user-images.githubusercontent.com/72132281/100517049-58cbda00-3188-11eb-882a-157a4db73677.png)

DenizShabani reviewed (commented):
Apart from the one problem mentioned already in the issue about the old components having the old behavior until routes are refactored i also noticed this. Dont know if this is supposed to be like this or not.
![image](https://user-images.githubusercontent.com/44246548/100521865-b3742e80-31a6-11eb-841d-736af70f1567.png)

airelawaleria reviewed (dismissed):
Tested all possible pages. From Server Administration everything is fine. From Course Management as well, except if you go to a specific course and then to "Exams", the breadcrumb of the course's name changes back to raw url, which is different from any other paths from the Course view like Exercises, Scores and so on. I attached two screenshots, whereas both are resulted from Course's view page but look differently. Since this PR covers only parts of Course Management, I assume it is not included in this pr, so approve.
![image](https://user-images.githubusercontent.com/64631786/100521899-08b04000-31a7-11eb-9364-640fc6536f6c.png)
![image](https://user-images.githubusercontent.com/64631786/100521905-1bc31000-31a7-11eb-939b-2cd06df4cb7e.png)

airelawaleria commented:
> Apart from the one problem mentioned already in the issue about the old components having the old behavior until routes are refactored i also noticed this. Dont know if this is supposed to be like this or not.
> ![image](https://user-images.githubusercontent.com/44246548/100521865-b3742e80-31a6-11eb-841d-736af70f1567.png)

I think there is no better way to identify a certain notification, since they don't have human-readable specific names.

marvinchimeka reviewed (dismissed):
tested on ts1. Beside the issue in the PR about the old components having the old behavior until routes are refactored, it's fine.

FrankeLukas commented:
> 
> 
> > Apart from the one problem mentioned already in the issue about the old components having the old behavior until routes are refactored i also noticed this. Dont know if this is supposed to be like this or not.
> > ![image](https://user-images.githubusercontent.com/44246548/100521865-b3742e80-31a6-11eb-841d-736af70f1567.png)
> 
> I think there is no better way to identify a certain notification, since they don't have human-readable specific names.

Yep, that is exactly the reason.

> Since this PR covers only parts of Course Management, I assume it is not included in this pr, so approve.

Yes, I didn't change the exam route in this PR.

Thanks for the fast reviews!

JohannesWeiss reviewed (dismissed):
Works as intended, with the mentioned exceptions.

FlorianStang reviewed (dismissed):
Tested on TS1, everything worked as intended, except of the mentioned exceptions (which retain the old behavior).

LudwigWeide reviewed (approved):
The breadcrumbs also jump back to the course ID when pressing on any of the fields within an exercise (marked in screenshot 1).
![bug_names_1](https://user-images.githubusercontent.com/52135815/100542629-6a29ea80-324b-11eb-98e1-f87c7a9757a0.PNG)
(schreenshot1)
This leads to:
![bug_names_2](https://user-images.githubusercontent.com/52135815/100542632-7150f880-324b-11eb-9ac1-97527edc5561.PNG)
(screenshot2)
I don´t know if this was already intended to be changed in this PR but just keep it in mind.

FrankeLukas commented:
> I don´t know if this was already intended to be changed in this PR but just keep it in mind.

Thanks for the heads up. I didn't change it in this PR so that is still expected. :)

EneaGore reviewed (dismissed):
Seems like a good start to improve paths and it went as described above.

krusche commented:
This is a PR with a lot of changes that might break something. We should really carefully test them before merging this PR

derLalla reviewed (changes requested):
good work! But within the Online Editor (using "Edit in Editor" for prog. ex) the breadcrumbs change to only two clickable items: the whole path and an id. See the first screenshot: 

![image](https://user-images.githubusercontent.com/33484318/100848107-f8d37d00-3480-11eb-9edc-d889ed7ce8e0.png)

Now, after clicking on that path breadcrumb the site renders white: 

![image](https://user-images.githubusercontent.com/33484318/100848171-0d177a00-3481-11eb-8f1d-932cd604982f.png)

FrankeLukas commented:
Rebased the merge conflicts away to keep this reviewable.

> good work! But within the Online Editor (using "Edit in Editor" for prog. ex) the breadcrumbs change to only two clickable items: the whole path and an id.

Good spot. That route didn't have its own routing but was only defined at the root level, hence I missed it. Fixed, thanks!

FrankeLukas commented:
Updated once again to resolve merge conflicts. (This time not via rebase as this seems to scare reviewers away I suppose.)

derLalla reviewed (dismissed):
Nice, the online editor is working now. But I guess the following is hard to solve: 
![image](https://user-images.githubusercontent.com/33484318/101134600-ef761c00-360a-11eb-8195-d7ca178fc20f.png)

When you click on `code-editor` the page goes blank, 
![image](https://user-images.githubusercontent.com/33484318/101134562-df5e3c80-360a-11eb-887d-cf5626f8ade2.png)
and the next step going to 4823 shows the correct overview page. 
![image](https://user-images.githubusercontent.com/33484318/101134681-0caaea80-360b-11eb-8787-f61cfd0a7721.png)

It would be nice to find a way to disable links which lead to blank pages.

airelawaleria reviewed (approved):
Everything works fine, except for the mentioned empty blanks

JohannesWeiss reviewed (commented):
Works as intended for the new parts (with the known exceptions).
Within the exercise scores and participations views the course id link does not work, if that is related?

![image](https://user-images.githubusercontent.com/37304775/101143501-4386fd80-3617-11eb-95f4-17a7a4a7c197.png)
E.g. clicking 4707 does not bring you to the exercise overview like in the editor page.

FrankeLukas commented:
> Within the exercise scores and participations views the course id link does not work, if that is related?

That is the same behaviour as on current `develop`. That will be fixed once I port that route to human-readable labels.

marvinchimeka reviewed (approved):
sorry I misclicked something. approve

tiffca reviewed (approved):
Tested this successfully on test server 1. The breadcrumb links that are updated work well. Noticed some breadcrumb updates are not yet implemented, as mentioned. Looks good!

simonzilker reviewed (approved):
Everything works as intended!

philgabr commented:
I just tested this as well. Basically everything works as intended except for the already mentioned aspects.

I spotted one additional exception. When you are on the test cases page, clicking on the breadcrumb "grading" does not change anything resp. you'll stay on that page:
![image](https://user-images.githubusercontent.com/73602874/101186699-b9a95580-3653-11eb-9dde-04bd5b10a382.png)

FrankeLukas commented:
> I spotted one additional exception. When you are on the test cases page, clicking on the breadcrumb "grading" does not change anything resp. you'll stay on that page:

Yes, this is also the behaviour on `develop` right now and will be resolved in a follow-up. :)

valeriebucher reviewed (dismissed):
the breadcrumb links look good!

ivanchimeno commented on src/main/webapp/app/lecture/lecture-update.component.ts:
```diff
@@ -40,9 +41,15 @@ export class LectureUpdateComponent implements OnInit {
+            this.course = course;
+            if (this.lecture) {
+                this.lecture.course = course;
+            }
```
I think that you can simplify the code from line 42 - 52 by using the [combineLatest](https://www.learnrxjs.io/learn-rxjs/operators/combination/combinelatest) operator from rxjs so that you wait for both observeables and then fill in the member variables.

FrankeLukas replied:
Ah, nice. I didn't know about that, added it in https://github.com/ls1intum/Artemis/pull/2474/commits/ea3340b2a9fdea9c9337ab1abb8b391f00d19b89.

ivanchimeno commented on src/test/javascript/spec/component/admin/user-management.component.spec.ts:
```diff
@@ -7,7 +7,7 @@ import { UserService } from 'app/core/user/user.service';
 import { AccountService } from 'app/core/auth/account.service';
 import { MockAccountService } from '../../helpers/mocks/service/mock-account.service';
 import { MockActivatedRoute, MockRouter } from '../../helpers/mocks/service/mock-route.service';
```
I think you missed the Codacy warning here :)

FrankeLukas replied:
I did, thanks for the hint. Fixed in https://github.com/ls1intum/Artemis/pull/2474/commits/7ba17b6c9528ef7291ce2cbd1d768938afda852c.

ivanchimeno reviewed (approved):
I added a couple comments which aren't really "request changes", other than that the code seems fine from my side. The `buildBreadcrumbs` function is a tad long but other than that :+1:

stefanwaldhauser reviewed (approved):
Awesome changed :) Code looks fine too 🚀

fde312 reviewed (approved):
Works and code looks also good as well. Good job! 🚀

airelawaleria reviewed (approved):
Everything works fine

JohannesWeiss reviewed (approved):
Tested in testing session. Re-Approve.

FuchsDominik reviewed (approved):
Looked at the code and tested. Works well 🚀

philippbauch reviewed (commented):
Tested during the testing session. Works well and the code also looks good 👍

philgabr reviewed (approved):
Tested in testing session. Everything works as intended, except for already mentioned aspects.

philippbauch reviewed (approved):
works well, code also looks good 👍

LudwigWeide reviewed (approved):
Retested on TS3. Works as intended.

tiffca reviewed (approved):
Places where it's implemented look good, links and breadcrumbs work as expected.
